### PR TITLE
ignore date expiration

### PR DESCRIPTION
**Pull request by michalfranc-form3**
implements https://github.com/tfsec/tfsec/issues/758

### new feature: expiration date for ignore
Adds in expiration date for ignore so that when adding ignore I can decide when I should be reminded about check that I am ignoring.

Current format
```
# tfsec:ignore:AWS006 exp:2221-12-02
```

### CODE REVIEW

**Comment by michalfranc-form3** [comment]
This is still WIP
but @owenrumney does it make any sense and is worth pursuing?

@vsetka any input ideas?

**Comment by owenrumney** [comment]
Can we go with no spaces when there is an expriy. something like 

```hcl
# tfsec:ignore:AWS069:exp:2021-12-02
```

**Comment by michal-franc** [comment]
ok @owenrumney `# tfsec:ignore:AWS069:exp:2021-12-02` will be now supported I ll just cover the edge cases.

**Comment by owenrumney** [comment]
Cool cool, sounds good

**Comment by michal-franc** [comment]
@owenrumney this is ready to be reviewed
The one thing requiring improvement is the way we should communicate to the user that the date format is probably incorrect and we were not able to parse it thus we cant use this ignore.
https://github.com/tfsec/tfsec/pull/762/files#diff-7d15dd8f6560094188d965b3619a075c1ee50bc7a80ce28374c9ea409d19957bR137

 Any ideas how to do it?
 I tried using `tml` but at this level it breaks unit tests as `tml` changes something in execution context.
 At the moment I have put `debug.log` in there.

File: internal/app/tfsec/scanner/scanner.go [hunk 1]
```diff
@@ -106,28 +110,43 @@ func (scanner *Scanner) checkRangeIgnored(id string, r block.Range, b block.Rang
...
+			if err != nil {
+				// if we can't parse the date then we don't want to ignore the range
+				debug.Log("Unable to parse exp date in ignore: '%s'. The date format is invalid. Supported format 'exp:yyyy-mm-dd'.", lineWithPotentialExp)
+				return false
+			}
+
+			currentTime := time.Now()
+			ignoreExpirationDateNotBreached := !currentTime.After(parsedDate)
+			if ignoreExpirationDateNotBreached {
+				debug.Log("Ignore ignored - expiration date not breached")
```
Comment (owenrumney) [inline internal/app/tfsec/scanner/scanner.go:144]: This logic should be reversed - if the ignore has expired then log to say `Ignore expired, check will be performed Filename: /file/path.tf:lineNo` The required info can be taken from `b`

**Review by owenrumney — CHANGES_REQUESTED** [review]

File: internal/app/tfsec/scanner/scanner.go [hunk 2]
```diff
@@ -106,28 +110,43 @@ func (scanner *Scanner) checkRangeIgnored(id string, r block.Range, b block.Rang
...
-	line = strings.TrimSpace(strings.ReplaceAll(strings.ReplaceAll(line, "//", ""), "#", ""))
-	segments := strings.Split(line, " ")
-	for _, segment := range segments {
-		if segment == ignoreAll || segment == ignoreCode {
-			return true
+	if foundValidIgnore {
+		lineWithPotentialExp := lines[lineValidIgnoreFound]
+		if indexExpFound := strings.Index(lineWithPotentialExp, "exp:"); indexExpFound > 0 {
+			debug.Log("Expiration date found on ignore '%s'", lineWithPotentialExp)
+			expDate := lineWithPotentialExp[indexExpFound:]
```
[the remark this answers is not in the document]
Comment (michal-franc) [reply internal/app/tfsec/scanner/scanner.go:132]: Found A way to do it without tokenising.

**Comment by michal-franc** [comment]
Reversed the logic:
Here is a snippet of how the debug log looks like.
```
[DEBUG][2021-06-24 19:07:06.300724775 +0200 CEST m=+0.001887818] Running rule for AWS006 on resource.aws_security_group_rule.my-rule (/home/michal-franc/Work/go-src/github.com/michal-franc/tfsec/bin/linux/test.tf)...
[DEBUG][2021-06-24 19:07:06.300753002 +0200 CEST m=+0.001916045] Expiration date found on ignore '# tfsec:ignore:AWS006:exp:2000-01-02 #tfsec:ignore:AWS018'
[DEBUG][2021-06-24 19:07:06.300761128 +0200 CEST m=+0.001924171] Ignore expired, check will be performed Filename: /home/michal-franc/Work/go-src/github.com/michal-franc/tfsec/bin/linux/test.tf:2
```

Also fixed the scenario with
```
#tfsec:ignore:AWS018:exp:2021-10-12 #tfsec:ignore:AWS006
```

Didn't have to make tokenization but instead as the checks are performed per rule we can actually check if line contains `<code>:exp` for example `AWS018:exp`.

File: internal/app/tfsec/scanner/scanner.go (hunk 1 above)
Comment (michal-franc) [reply internal/app/tfsec/scanner/scanner.go]: done example in the comment below.

**Review by owenrumney — DISMISSED** [review]

**Comment by owenrumney** [comment]
Thanks @michal-franc  - just need to sync your branch then we can merge

**Comment by michal-franc** [comment]
@owenrumney I have just rebased it.

**Comment by michal-franc** [comment]
Updating docs now and it is good to go.

**Review by owenrumney — APPROVED** [review]

**Pull request merged**